RPCv04 Remove BroadcastTx

rianhughes opened the pull request:
closes: https://github.com/NethermindEth/starknet.go/issues/278
closes: https://github.com/NethermindEth/starknet.go/issues/230

Note: Tests are fixed in previous rpcv04 PRs
Note: Will have a merge conflict with previous rpcv04 PRs. Should be merged last.

rianhughes committed:
RPCv04 Remove BroadcastTx

rianhughes committed:
merge main

rianhughes committed:
merge main

rianhughes committed:
merge main

rianhughes committed:
merge main

cicr99 commented on account.go:
```diff
@@ -351,13 +351,11 @@ func (account *Account) prepFunctionInvokeRPC(ctx context.Context, messageType s
-				Signature: []*felt.Felt{s1Felt, s2Felt},
-				Nonce:     nonceFelt,
-			},
+		return &rpc.InvokeTxnV1{
```
Shouldn't the `type` also be set here?

rianhughes replied:
Probably, although I think this is out of scope of this PR for two reasons 1) the goal was to replace broacastInvokeTxn with InvokeTxn types (which this PR does) not fix the logic in account.go, and 2) this account file is broken, and will be replaced with the new accounts package.

cicr99 replied:
I thought to point it out since `Type` was a required field for the `InvokeTxnV1` type, but you are right, this file is already broken and will be replaced soon

cicr99 commented on account.go:
```diff
@@ -451,14 +449,13 @@ func (account *Account) EstimateFee(ctx context.Context, calls []types.FunctionC
+				Version:       rpc.TransactionV1,
+				Signature:     call.Signature,
+				Nonce:         call.Nonce,
+				Type:          "INVOKE",
```
Could we use `TransactionType_Invoke` instead? (https://github.com/NethermindEth/starknet.go/blob/a3a745a7309046e1fdbd1d649c867c58759617ca/rpc/types_transaction_receipt.go#L43)

rianhughes replied:
Same comment as above

cicr99 commented on account.go:
```diff
@@ -501,14 +498,12 @@ func (account *Account) Execute(ctx context.Context, calls []types.FunctionCall,
+				Version:       rpc.TransactionV1,
+				Signature:     call.Signature,
+				Nonce:         call.Nonce,
+				Type:          "INVOKE",
```
Same here about using `TransactionType_Invoke`

rianhughes replied:
Same comment as above

cicr99 commented on contracts/contractsv02.go:
```diff
@@ -18,6 +17,7 @@ import (
 type RPCProvider rpc.Provider
 
 func (p *RPCProvider) declareAndWaitWithWallet(ctx context.Context, compiledClass []byte) (*DeclareOutput, error) {
+	panic("This needs updated - Declare transactions have changed significantly over the past few rpc-spec updates")
```
I know after this line the rest of the code of the function is unreachable but can we comment it out same as in the other methods?

cicr99 commented on rpc/types_transaction.go:
```diff
@@ -104,7 +108,12 @@ type DeclareTxnV0 struct {
 
 type DeclareTxnV1 struct {
-	CommonTransaction
+	TransactionHash *felt.Felt         `json:"transaction_hash,omitempty"`
```
Why `TransactionHash`? I couldn't find it in the [spec](https://github.com/starkware-libs/starknet-specs/blob/6fadca1f1abf48523c0aa799a10061f40f973468/api/starknet_api_openrpc.json#L1687)

cicr99 replied:
I've seen most of the transaction types have this property as well, but there is the same question

rianhughes replied:
Looks like it found it's way back in after merging main a couple of times. It wasn't there in the original PR, but you're right that it should be removed.

cicr99 commented on rpc/types_transaction.go:
```diff
@@ -104,7 +108,12 @@ type DeclareTxnV0 struct {
+	Type            TransactionType    `json:"type"`
 
 	// ClassHash the hash of the declared class
 	ClassHash *felt.Felt `json:"class_hash"`
```
Should we add _"omitempty"_ for `ClassHash` and `DeprecatedContractClass`? One of them is required, but you shouldn't get both at the same time, so there's always one of them that will be empty

cicr99 commented on rpc/write.go:
```diff
@@ -4,13 +4,11 @@ import (
-type BroadcastedInvokeTransaction interface{}
-
-func (provider *Provider) AddInvokeTransaction(ctx context.Context, broadcastedInvoke BroadcastedInvokeTransaction) (*AddInvokeTransactionResponse, error) {
+func (provider *Provider) AddInvokeTransaction(ctx context.Context, invokeTxn AddInvokeTxnInput) (*AddInvokeTransactionResponse, error) {
```
Why use `AddInvokeTxnInput` instead of just `InvokeTxnV1`? (See [here](https://github.com/starkware-libs/starknet-specs/blob/6fadca1f1abf48523c0aa799a10061f40f973468/api/starknet_write_api.json#L19))

rianhughes replied:
It's because we used to allow InvokeTxnV0 and InvokeTxnV1 for the rpcv03 spec, which took [broadcastInvokeTxn](https://github.com/starkware-libs/starknet-specs/blob/v0.3.0/api/starknet_api_openrpc.json#L1902), I'll update it though

cicr99 commented on rpc/types_transaction_response.go:
```diff
@@ -14,6 +14,12 @@ type AddDeployAccountTransactionResponse struct {
+type AddDeployTransactionResponse struct {
+	TransactionHash *felt.Felt `json:"transaction_hash"`
+	ContractAddress *felt.Felt `json:"contract_address"`
+}
```
This type looks exactly the same as `AddDeployAccountTransactionResponse`, and I don't see it being used anywhere on the code. What's the point?

Also both types have the same comment on the top (probably because of copy-pasting), if they are meant for different things it 'll be good to change the comments

rianhughes replied:
This is no longer needed, I'll remove it

cicr99 commented on rpc/contract.go:
```diff
@@ -111,15 +111,8 @@ func (provider *Provider) Nonce(ctx context.Context, blockID BlockID, contractAd
-	// 	tx.EntryPointSelector = fmt.Sprintf("0x%x", types.GetSelectorFromName(tx.EntryPointSelector))
-	// 	request = tx
-	// }
+// EstimateFee estimates the fee for a given StarkNet transaction.
```
Use **Starknet** instead of **StarkNet**

cicr99 commented on contracts/contractsv02.go:
```diff
@@ -27,14 +27,13 @@ func (p *RPCProvider) declareAndWaitWithWallet(ctx context.Context, compiledClas
+	tx, err := provider.AddDeclareTransaction(ctx, rpc.DeclareTxnV1{
+		Type:    "DECLARE",
+		MaxFee:  new(felt.Felt).SetUint64(10000),
+		Version: "0x01",
```
I would use [TransactionType_Declare](https://github.com/NethermindEth/starknet.go/blob/a3a745a7309046e1fdbd1d649c867c58759617ca/rpc/types_transaction_receipt.go#L40) for the `Type` and [TransactionV1](https://github.com/NethermindEth/starknet.go/blob/a3a745a7309046e1fdbd1d649c867c58759617ca/rpc/types_transaction.go#L292C2-L292C15) for the `Version`

rianhughes replied:
Same comment as above

rianhughes committed:
Address comments + add BlockTransaction type

The spec requires the transaction types to not contain a transaction
hash field. However, the block contains a list of transactions types
which do contain a transaction hash. To solve this, I introduced a new
type called BlockTransactions, BlockInvokeTxnV1, etc to account for
these distinct use cases.

cicr99 commented on rpc/block_test.go:
```diff
@@ -220,34 +220,31 @@ func TestBlockWithTxsAndInvokeTXNV0(t *testing.T) {
-						},
-						Nonce: nil,
+		Transactions: []BlockTransaction{
+			BlockInvokeTxnV0{
```
Shouldn't this also have a `TransactionHash` field set? I assume you should be able to use the same one that is being removed

cicr99 commented on rpc/block_test.go:
```diff
@@ -306,19 +303,17 @@ func TestBlockWithTxsAndInvokeTXNV0(t *testing.T) {
 				t.Fatal("expected invoke v0 transaction")
 			}
-
-			require.Equal(t, invokeV0Want.Hash(), invokeV0Block.Hash(), "expected equal hash")
```
You could also modify this one according to previous comment about `TransactionHash`

cicr99 commented on rpc/block_test.go:
```diff
@@ -348,11 +343,10 @@ func TestBlockWithTxsAndDeployOrDeclare(t *testing.T) {
-				ClassHash:       utils.TestHexToFelt(t, "0x1ca349f9721a2bf05012bb475b404313c497ca7d6d5f80c03e98ff31e9867f5"),
-				DeployTransactionProperties: DeployTransactionProperties{
+		Transactions: []BlockTransaction{
+			BlockDeployTxn{
```
Same thing here about `TransactionHash`

cicr99 commented on rpc/block_test.go:
```diff
@@ -376,20 +370,17 @@ func TestBlockWithTxsAndDeployOrDeclare(t *testing.T) {
-						Nonce:     utils.TestHexToFelt(t, "0x333"),
-					},
+		Transactions: []BlockTransaction{
+			BlockDeclareTxnV1{
```
Same thing here about `TransactionHash`

cicr99 commented on rpc/block_test.go:
```diff
@@ -404,21 +395,18 @@ func TestBlockWithTxsAndDeployOrDeclare(t *testing.T) {
-						Nonce:     utils.TestHexToFelt(t, "0xd"),
-					},
+		Transactions: []BlockTransaction{
+			BlockDeclareTxnV2{
```
Same thing here about `TransactionHash`

cicr99 commented on rpc/types_transaction.go:
```diff
@@ -304,53 +268,19 @@ func (v *TransactionVersion) BigInt() (*big.Int, error) {
 
-var _ BroadcastedDeclareTransaction = BroadcastedDeclareTransactionV1{}
-var _ BroadcastedDeclareTransaction = BroadcastedDeclareTransactionV2{}
+type AddDeployAccountTxnInput interface{}
```
Why do you need this one? There is only one `DeployAccountTxn`. Also, this is only being used in `AddDeployAccountTransaction` method which could just receive a `DeployAccountTxn` as an input parameter ([spec](https://github.com/starkware-libs/starknet-specs/blob/6fadca1f1abf48523c0aa799a10061f40f973468/api/starknet_write_api.json#L152C55-L152C73))

rianhughes committed:
address carment comments txhash

rianhughes committed:
include template
